pims/base_frames.py

```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+    """
+    def clear_dims(self):
+        self._dims = []
+
+    def add_dim(self, name, size, aggregate=False, iterate=True,
+                default=0):
+        new_dim = FramesDimension(name, size, aggregate, iterate, default)
+        self._dims += [new_dim]
+
+    def __len__(self):
+        return np.prod([d.size for d in self._dims if d.iterate])
+
+    @abstractproperty
+    def frame_shape_2D(self, **ind):
+        """ This property should return the shape of a single 2D image. """
+        pass
+
+    @property
+    def frame_shape(self):
+        """ Returns the shape of the frame as returned by get_frame. """
+        shape = [d.size for d in self._dims if d.aggregate]
+        return tuple(shape + list(self.frame_shape_2D))
+
+    @property
+    def dims(self):
+        """ Returns a dictionary of all FramesDimension objects. """
+        return {d.name: d for d in self._dims}
+
+    @property
+    def ndim(self):
+        """ Returns the number of dimensions, including x and y. """
+        return len(self._dims) + 2
+
+    @property
+    def sizes(self):
+        """ Returns a dict of all dimension sizes, including x and y. """
+        result = {d.name: d.size for d in self._dims}
+        result['y'], result['x'] = self.frame_shape_2D
+        return result
+
+    @property
+    def aggregate(self):
+        """ This determines which dimensions will be aggregated into one Frame.
+        The ndarray that is returned by get_frame has the same dimension order
+        as the order of aggregate. """
+        return [d.name for d in self._dims if d.aggregate]
+
+    @aggregate.setter
+    def aggregate(self, value):
+        for dim_aggregate in value:
+            if dim_aggregate.lower() in ['x', 'y']:
+                raise ValueError('Aggregate does not take dimensions x or y.')
+            if dim_aggregate not in self.dims:
+                raise ValueError(('Dimension named ''{}'' does not exist ' +
+                                  'in this image.').format(dim_aggregate))
+
+        new_dims_aggr = [0] * len(value)
+        new_dims_rest = []
+        for dim in self._dims:
+            for n_aggr, dim_aggregate in enumerate(value):
+                if dim.name == dim_aggregate:
+                    new_dims_aggr[n_aggr] = dim
+                    dim.aggregate = True
+                    break
+            else:
+                new_dims_rest.append(dim)
+                dim.aggregate = False
+        self._dims = new_dims_rest + new_dims_aggr
+
+    @property
+    def iterate(self):
+        """ This determines which dimensions will be iterated over by the
+        FramesSequence. The last element will iterate fastest. """
+        return [d.name for d in self._dims if d.iterate]
+
+    @iterate.setter
+    def iterate(self, value):
+        for dim_iterate in value:
+            if dim_iterate.lower() in ['x', 'y']:
+                raise ValueError('Iterate does not take dimensions x or y.')
```

Comment:
I guess I took `ValueError` because the user is trying to set a non-existing dimension to `iterate`